README
NAME
    Perl::Critic - Critique Perl source code for best-practices.

SYNOPSIS
        use Perl::Critic;
        my $file = shift;
        my $critic = Perl::Critic->new();
        my @violations = $critic->critique($file);
        print @violations;

DESCRIPTION
    Perl::Critic is an extensible framework for creating and applying coding
    standards to Perl source code. Essentially, it is a static source code
    analysis engine. Perl::Critic is distributed with a number of
    Perl::Critic::Policy modules that attempt to enforce various coding
    guidelines. Most Policy modules are based on Damian Conway's book Perl
    Best Practices. However, Perl::Critic is not limited to PBP and will
    even support Policies that contradict Conway. You can enable, disable,
    and customize those Polices through the Perl::Critic interface. You can
    also create new Policy modules that suit your own tastes.

    For a command-line interface to Perl::Critic, see the documentation for
    perlcritic. If you want to integrate Perl::Critic with your build
    process, Test::Perl::Critic provides an interface that is suitable for
    test scripts. Also, Test::Perl::Critic::Progressive is useful for
    gradually applying coding standards to legacy code. For the ultimate
    convenience (at the expense of some flexibility) see the criticism
    pragma.

    Win32 and ActivePerl users can find PPM distributions of Perl::Critic at
    <http://theoryx5.uwinnipeg.ca/ppms/>.

    If you'd like to try Perl::Critic without installing anything, there is
    a web-service available at <http://perlcritic.com>. The web-service does
    not yet support all the configuration features that are available in the
    native Perl::Critic API, but it should give you a good idea of what it
    does. You can also invoke the perlcritic web-service from the
    command-line by doing an HTTP-post, such as one of these:

        $> POST http://perlcritic.com/perl/critic.pl < MyModule.pm
        $> lwp-request -m POST http://perlcritic.com/perl/critic.pl < MyModule.pm
        $> wget -q -O - --post-file=MyModule.pm http://perlcritic.com/perl/critic.pl

    Please note that the perlcritic web-service is still alpha code. The URL
    and interface to the service are subject to change.

    Also, the Perl Development Kit (PDK 8.0) from ActiveState includes a
    very slick graphical interface to Perl-Critic. For details, go to
    <http://www.activestate.com/perl_dev_kit>

INTERFACE SUPPORT
    This is considered to be a public class. Any changes to its interface
    will go through a deprecation cycle.

CONSTRUCTOR
    "new( [ -profile => $FILE, -severity => $N, -theme => $string, -include
    => \@PATTERNS, -exclude => \@PATTERNS, -top => $N, -only => $B,
    -profile-strictness => $PROFILE_STRICTNESS_{WARN|FATAL|QUIET}, -force =>
    $B, -verbose => $N ], -color => $B, -pager => $string, -criticism-fatal
    => $B)"
    "new()"
        Returns a reference to a new Perl::Critic object. Most arguments are
        just passed directly into Perl::Critic::Config, but I have described
        them here as well. The default value for all arguments can be
        defined in your .perlcriticrc file. See the "CONFIGURATION" section
        for more information about that. All arguments are optional
        key-value pairs as follows:

        -profile is a path to a configuration file. If $FILE is not defined,
        Perl::Critic::Config attempts to find a .perlcriticrc configuration
        file in the current directory, and then in your home directory.
        Alternatively, you can set the "PERLCRITIC" environment variable to
        point to a file in another location. If a configuration file can't
        be found, or if $FILE is an empty string, then all Policies will be
        loaded with their default configuration. See "CONFIGURATION" for
        more information.

        -severity is the minimum severity level. Only Policy modules that
        have a severity greater than $N will be applied. Severity values are
        integers ranging from 1 (least severe violations) to 5 (most severe
        violations). The default is 5. For a given "-profile", decreasing
        the "-severity" will usually reveal more Policy violations. You can
        set the default value for this option in your .perlcriticrc file.
        Users can redefine the severity level for any Policy in their
        .perlcriticrc file. See "CONFIGURATION" for more information.

        If it is difficult for you to remember whether severity "5" is the
        most or least restrictive level, then you can use one of these named
        values:

            SEVERITY NAME   ...is equivalent to...   SEVERITY NUMBER
            --------------------------------------------------------
            -severity => 'gentle'                     -severity => 5
            -severity => 'stern'                      -severity => 4
            -severity => 'harsh'                      -severity => 3
            -severity => 'cruel'                      -severity => 2
            -severity => 'brutal'                     -severity => 1

        -theme is special expression that determines which Policies to apply
        based on their respective themes. For example, the following would
        load only Policies that have a 'bugs' AND 'pbp' theme:

          my $critic = Perl::Critic->new( -theme => 'bugs && pbp' );

        Unless the "-severity" option is explicitly given, setting "-theme"
        silently causes the "-severity" to be set to 1. You can set the
        default value for this option in your .perlcriticrc file. See the
        "POLICY THEMES" section for more information about themes.

        -include is a reference to a list of string @PATTERNS. Policy
        modules that match at least one "m/$PATTERN/ixms" will always be
        loaded, irrespective of all other settings. For example:

            my $critic = Perl::Critic->new(-include => ['layout'] -severity => 4);

        This would cause Perl::Critic to apply all the "CodeLayout::*"
        Policy modules even though they have a severity level that is less
        than 4. You can set the default value for this option in your
        .perlcriticrc file. You can also use "-include" in conjunction with
        the "-exclude" option. Note that "-exclude" takes precedence over
        "-include" when a Policy matches both patterns.

        -exclude is a reference to a list of string @PATTERNS. Policy
        modules that match at least one "m/$PATTERN/ixms" will not be
        loaded, irrespective of all other settings. For example:

            my $critic = Perl::Critic->new(-exclude => ['strict'] -severity => 1);

        This would cause Perl::Critic to not apply the "RequireUseStrict"
        and "ProhibitNoStrict" Policy modules even though they have a
        severity level that is greater than 1. You can set the default value
        for this option in your .perlcriticrc file. You can also use
        "-exclude" in conjunction with the "-include" option. Note that
        "-exclude" takes precedence over "-include" when a Policy matches
        both patterns.

        -single-policy is a string "PATTERN". Only one policy that matches
        "m/$PATTERN/ixms" will be used. Policies that do not match will be
        excluded. This option has precedence over the "-severity", "-theme",
        "-include", "-exclude", and "-only" options. You can set the default
        value for this option in your .perlcriticrc file.

        -top is the maximum number of Violations to return when ranked by
        their severity levels. This must be a positive integer. Violations
        are still returned in the order that they occur within the file.
        Unless the "-severity" option is explicitly given, setting "-top"
        silently causes the "-severity" to be set to 1. You can set the
        default value for this option in your .perlcriticrc file.

        -only is a boolean value. If set to a true value, Perl::Critic will
        only choose from Policies that are mentioned in the user's profile.
        If set to a false value (which is the default), then Perl::Critic
        chooses from all the Policies that it finds at your site. You can
        set the default value for this option in your .perlcriticrc file.

        -profile-strictness is an enumerated value, one of
        "$PROFILE_STRICTNESS_WARN" in Perl::Critic::Utils::Constants (the
        default), "$PROFILE_STRICTNESS_FATAL" in
        Perl::Critic::Utils::Constants, and "$PROFILE_STRICTNESS_QUIET" in

Changes
[1.105] Released on 2009-09-07

    Bug Fixes:
    * Variables::ProhibitPunctuationVars would complain about "%-" appearing
      anywhere in a string.  RT #49016

    Policy Changes:
    * InputOutput::RequireCheckedSyscalls now complains about unchecked "say"
      by default.  RT #37487

[1.104] Released on 2009-08-23

    This release is dedicated to Tom Wyant in appreciation of the amount of
    effort he put into the enhancements and bug fixes in this release, for
    having the patience to wait for the amount of time that it took to get
    them out, and for overall awesomeness.  Thank you, Tom.

    New Policies:
      Objects::ProhibitIndirectSyntax
      ValuesAndExpressions::ProhibitComplexVersion
      ValuesAndExpressions::RequireConstantVersion

    New Optional Requirement:
    * Email::Address, if you want
      ValuesAndExpressions::ProhibitInterpolationOfLiterals to properly ignore
      email addresses.

    New Features:
    * Perlcritic will list the names of files with violations if given the
      --files-with-violations option, or the names of files without
      violations if given the --files-without-violations options. These
      have synonyms -l and -L respectively.
    * Perlcritic has a new --list-enabled option, which lists the Policies
      that will be enforced, given the current configuration.  This is
      useful if you've written a complex command-line or modified your
      .perlcriticrc file and you want to see which Policies *would*
      be used with the current configuration, if you were actually going
      to critque a file with it.
    * Perl::Critic::Violation now takes #line directives into account in the
      %F, %f, and %l formats.  You can get the old values via the new %G, %g,
      and %L formats.

    Policy Changes:
    * CodeLayout::ProhibitParensWithBuiltins was complaining in certain cases
      where parentheses are required due to operator precedence.  RT #46862.
    * ControlStructures::ProhibitMutatingListFunctions no longer complains
      about uses of tr/// that don't modify the operand.  Reported by EDAVIS,
      RT #44515.
    * Miscellanea::RequireRcsKeywords now accepts "qw$Keyword: ...$".  RT
      #45196.
    * Modules::RequireFilenameMatchesPackage now respects logical filenames
      defined by the "#line" directives.  This allows the Policy to work
      properly with IDEs and code generators.
    * NamingConventions::Capitalization now allows fully qualified subroutine
      declarations ( e.g. "sub Foo::Bar::baz {...}" ).  However, the
      non-package part of the subroutine name must still conform to whatever
      capitalization rule you have chosen.
    * RegularExpressions::ProhibitCaptureWithoutTest no longer complains if
      the regex is followed by an "or die" or similar.  Reported by EDAVIS,
      RT #36081.
    * RegularExpressions::ProhibitComplexRegexes no longer counts variable
      substitutions in the length.   Reported by EDAVIS, RT #36098.
    * RegularExpressions::ProhibitUnusedCapture now considers the body of
      while loops and not just their condition.  Reported by EDAVIS, RT
      #38942.
    * ValuesAndExpressions::ProhibitVersionStrings was getting confused by
      comments.  Reported by Kevin Ryde,  RT #44986.
    * ValuesAndExpressions::RequireInterpolationOfMetachars now allows sigils
      in the arguments to "use vars".  Contributed by Kevin Ryde, RT #47318.
    * ValuesAndExpressions::RequireInterpolationOfMetachars now properly
      ignores email addesses, if you have Email::Address installed.  Inspired
      by the Kevin Ryde contribution in RT #47318.
    * Variables::ProhibitPunctuationVars gained the ability to look inside
      interpolated strings.  Doing this correctly is challenging and things
      may not work out right; how the policy does this can be controlled via
      the new "string_mode" option.  Contributed by Edgar Whipple
      <perlmonk at misterwhipple dot com>.
    * Variables::ProhibitPunctuationVars now ignores $] by default since there
      is no English.pm equivalent.

    Other Bug Fixes:
    * Perl::Critic::Utils::parse_arg_list() was slurping up the "or die ..."
      portion of "open my $foo, 'somefile' or die ...", causing

lib/Perl/Critic/Command.pm
##############################################################################
#      $URL: http://perlcritic.tigris.org/svn/perlcritic/branches/Perl-Critic-backlog/lib/Perl/Critic/Command.pm $
#     $Date: 2009-09-07 16:19:21 -0500 (Mon, 07 Sep 2009) $
#   $Author: clonezone $
# $Revision: 3629 $
##############################################################################

package Perl::Critic::Command;

use 5.006001;
use strict;
use warnings;

use English qw< -no_match_vars >;
use Readonly;

use Getopt::Long qw< GetOptions >;
use List::Util qw< first max >;
use Pod::Usage qw< pod2usage >;

use Perl::Critic::Exception::Parse ();
use Perl::Critic::Utils qw<
    :characters :severities policy_short_name
    $DEFAULT_VERBOSITY $DEFAULT_VERBOSITY_WITH_FILE_NAME
>;
use Perl::Critic::Violation qw<>;

#-----------------------------------------------------------------------------

our $VERSION = '1.105';

#-----------------------------------------------------------------------------

use base 'Exporter';

Readonly::Array our @EXPORT_OK => qw< run >;

Readonly::Hash our %EXPORT_TAGS => (
    all             => [ @EXPORT_OK ],
);

#-----------------------------------------------------------------------------

Readonly::Scalar my $DEFAULT_VIOLATIONS_FOR_TOP => 20;

Readonly::Scalar my $EXIT_SUCCESS           => 0;
Readonly::Scalar my $EXIT_NO_FILES          => 1;
Readonly::Scalar my $EXIT_HAD_VIOLATIONS    => 2;
Readonly::Scalar my $EXIT_HAD_FILE_PROBLEMS => 3;

#-----------------------------------------------------------------------------

my @files = ();
my $critic = undef;
my $output = \*STDOUT;

#-----------------------------------------------------------------------------

sub _out {
    my @lines = @_;
    return print {$output} @lines;
}

#-----------------------------------------------------------------------------

sub run {
    my %options    = _get_options();
    @files         = _get_input(@ARGV);

    my ($violations, $had_error_in_file) = _critique(\%options, @files);

    return $EXIT_HAD_FILE_PROBLEMS  if $had_error_in_file;
    return $EXIT_NO_FILES           if not defined $violations;
    return $EXIT_HAD_VIOLATIONS     if $violations;

    return $EXIT_SUCCESS;
}

#-----------------------------------------------------------------------------

sub _get_options {

    my %opts = _parse_command_line();
    _dispatch_special_requests( %opts );
    _validate_options( %opts );

    # Convert severity shortcut options.  If multiple shortcuts
    # are given, the lowest one wins.  If an explicit --severity
    # option has been given, then the shortcuts are ignored. The
    # @SEVERITY_NAMES variable is exported by Perl::Critic::Utils.
    $opts{-severity} ||= first { exists $opts{"-$_"} } @SEVERITY_NAMES;
    $opts{-severity} ||= first { exists $opts{"-$_"} } ($SEVERITY_LOWEST ..  $SEVERITY_HIGHEST);


    # If --top is specified, default the severity level to 1, unless an
    # explicit severity is defined.  This provides us flexibility to
    # report top-offenders across just some or all of the severity levels.
    # We also default the --top count to twenty if none is given
    if ( exists $opts{-top} ) {
        $opts{-severity} ||= 1;
        $opts{-top} ||= $DEFAULT_VIOLATIONS_FOR_TOP;
    }

    #Override profile, if --noprofile is specified
    if ( exists $opts{-noprofile} ) {
        $opts{-profile} = $EMPTY;
    }

    return %opts;
}

#-----------------------------------------------------------------------------

sub _parse_command_line {
    my %opts;
    my @opt_specs = _get_option_specification();
    Getopt::Long::Configure('no_ignore_case');
    GetOptions( \%opts, @opt_specs ) || pod2usage();           #Exits

    # I've adopted the convention of using key-value pairs for
    # arguments to most functions.  And to increase legibility,
    # I have also adopted the familiar command-line practice
    # of denoting argument names with a leading dash (-).
    my %dashed_opts = map { ( "-$_" => $opts{$_} ) } keys %opts;
    return %dashed_opts;
}

#-----------------------------------------------------------------------------

sub _dispatch_special_requests {
    my (%opts) = @_;
    if ( $opts{-help}            ) { pod2usage( -verbose => 0 )    }  # Exits
    if ( $opts{-options}         ) { pod2usage( -verbose => 1 )    }  # Exits
    if ( $opts{-man}             ) { pod2usage( -verbose => 2 )    }  # Exits
    if ( $opts{-version}         ) { _display_version()            }  # Exits
    if ( $opts{-list}            ) { _render_all_policy_listing()  }  # Exits
    if ( $opts{'-list-enabled'}  ) { _render_policy_listing(%opts) }  # Exits
    if ( $opts{'-list-themes'}   ) { _render_theme_listing()       }  # Exits
    if ( $opts{'-profile-proto'} ) { _render_profile_prototype()   }  # Exits
    if ( $opts{-doc}             ) { _render_policy_docs( %opts )  }  # Exits
    return 1;
}

#-----------------------------------------------------------------------------

sub _validate_options {
    my (%opts) = @_;
    my $msg = $EMPTY;


    if ( $opts{-noprofile} && $opts{-profile} ) {
        $msg .= qq{Warning: Cannot use -noprofile with -profile option.\n};
    }

    if ( $opts{-verbose} && $opts{-verbose} !~ m{(?: \d+ | %[mfFlcCedrpPs] )}xms) {
        $msg .= qq<Warning: --verbose arg "$opts{-verbose}" looks odd.  >;
        $msg .= qq<Perhaps you meant to say "--verbose 3 $opts{-verbose}."\n>;
    }

    if ( exists $opts{-top} && $opts{-top} < 0 ) {
        $msg .= qq<Warning: --top argument "$opts{-top}" is negative.  >;
        $msg .= qq<Perhaps you meant to say "$opts{-top} --top".\n>;
    }

    if (
            exists $opts{-severity}
        &&  (
                    $opts{-severity} < $SEVERITY_LOWEST
                ||  $opts{-severity} > $SEVERITY_HIGHEST
            )
    ) {
        $msg .= qq<Warning: --severity arg "$opts{-severity}" out of range.  >;
        $msg .= qq<Severities range from "$SEVERITY_LOWEST" (lowest) to >;
        $msg .= qq<"$SEVERITY_HIGHEST" (highest).\n>;
    }


    if ( $msg ) {
        pod2usage( -exitstatus => 1, -message => $msg, -verbose => 0); #Exits
    }


    return 1;
}

#-----------------------------------------------------------------------------

sub _get_input {

    my @args = @_;

    if ( !@args || (@args == 1 && $args[0] eq q{-}) )  {

        # Reading code from STDIN.  All the code is slurped into
        # a string.  PPI will barf if the string is just whitespace.
        my $code_string = do { local $RS = undef; <STDIN> };

        # Notice if STDIN was closed (pipe error, etc)
        if ( ! defined $code_string ) {
            $code_string = $EMPTY;
        }

        $code_string =~ m{ \S+ }xms || die qq{Nothing to critique.\n};
        return \$code_string;    #Convert to SCALAR ref for PPI
    }
    else {

        # Test to make sure all the specified files or directories
        # actually exist.  If any one of them is bogus, then die.
        if ( my $nonexistent = first { ! -e $_ } @args ) {
            my $msg = qq{No such file or directory: '$nonexistent'};
            pod2usage( -exitstatus => 1, -message => $msg, -verbose => 0);
        }

        # Reading code from files or dirs.  If argument is a file,
        # then we process it as-is (even though it may not actually
        # be Perl code).  If argument is a directory, recursively
        # search the directory for files that look like Perl code.
        return map { -d $_ ? Perl::Critic::Utils::all_perl_files($_) : $_ } @args;
    }
}

#------------------------------------------------------------------------------

sub _critique {

    my ( $opts_ref, @files_to_critique ) = @_;
    @files_to_critique || die "No perl files were found.\n";

    # Perl::Critic has lots of dependencies, so loading is delayed
    # until it is really needed.  This hack reduces startup time for
    # doing other things like getting the version number or dumping
    # the man page. Arguably, those things are pretty rare, but hey,
    # why not save a few seconds if you can.

    require Perl::Critic;
    $critic = Perl::Critic->new( %{$opts_ref} );
    $critic->policies() || die "No policies selected.\n";

    _set_up_pager($critic->config()->pager());

    my $number_of_violations = undef;
    my $had_error_in_file = 0;

    for my $file (@files_to_critique) {

        eval {
            my @violations = $critic->critique($file);
            $number_of_violations += scalar @violations;

            if (not $opts_ref->{'-statistics-only'}) {
                _render_report( $file, $opts_ref, @violations )
            }
            1;
        }
        or do {
            if ( my $exception = Perl::Critic::Exception::Parse->caught() ) {
                $had_error_in_file = 1;
                warn qq<Problem while critiquing "$file": $EVAL_ERROR\n>;
            }
            elsif ($EVAL_ERROR) {
                # P::C::Exception::Fatal includes the stack trace in its
                # stringification.
                die qq<Fatal error while critiquing "$file": $EVAL_ERROR\n>;
            }
            else {
                die qq<Fatal error while critiquing "$file". Unfortunately, >,
                    q<$@/$EVAL_ERROR >, ## no critic (RequireInterpolationOfMetachars)
                    qq<is empty, so the reason can't be shown.\n>;
            }
        }
    }

    if ( $opts_ref->{-statistics} or $opts_ref->{'-statistics-only'} ) {
        my $stats = $critic->statistics();
        _report_statistics( $opts_ref, $stats );
    }

    return $number_of_violations, $had_error_in_file;
}

#------------------------------------------------------------------------------

sub _render_report {
    my ( $file, $opts_ref, @violations ) = @_;

    # Only report the files, if asked.
    my $number_of_violations = scalar @violations;
    if ( $opts_ref->{'-files-with-violations'} ||
        $opts_ref->{'-files-without-violations'} ) {
        not ref $file
            and $opts_ref->{$number_of_violations ? '-files-with-violations' :
            '-files-without-violations'}
            and _out "$file\n";
        return $number_of_violations;
    }

    # Only report the number of violations, if asked.
    if( $opts_ref->{-count} ){
        ref $file || _out "$file: ";
        _out "$number_of_violations\n";
        return $number_of_violations;
    }

    # Hail all-clear unless we should shut up.
    if( !@violations && !$opts_ref->{-quiet} ) {
        ref $file || _out "$file ";
        _out "source OK\n";
        return 0;
    }

    # Otherwise, format and print violations
    my $verbosity = $critic->config->verbose();
    # $verbosity can be numeric or string, so use "eq" for comparison;
    $verbosity =
        ($verbosity eq $DEFAULT_VERBOSITY && @files > 1)
            ? $DEFAULT_VERBOSITY_WITH_FILE_NAME
            : $verbosity;
    my $fmt = Perl::Critic::Utils::verbosity_to_format( $verbosity );
    if (not -f $file) { $fmt =~ s< \%[fF] ><STDIN>xms; } #HACK!
    Perl::Critic::Violation::set_format( $fmt );

    my $color = $critic->config->color();
    _out $color ? _colorize_by_severity(@violations) : @violations;

    return $number_of_violations;
}

#-----------------------------------------------------------------------------

sub _set_up_pager {
    my ($pager_command) = @_;
    return if not $pager_command;
    return if not _at_tty();

    open my $pager, q<|->, $pager_command  ## no critic (InputOutput::RequireBriefOpen)
        or die qq<Unable to pipe to pager "$pager_command": $ERRNO\n>;

    $output = $pager;

    return;
}

#-----------------------------------------------------------------------------

sub _report_statistics {
    my ($opts_ref, $statistics) = @_;

    if (
            not $opts_ref->{'-statistics-only'}
        and (
                $statistics->total_violations()
            or  not $opts_ref->{-quiet} and $statistics->modules()
        )
    ) {
        _out "\n"; # There's prior output that we want to separate from.
    }

    my $files = _commaify($statistics->modules());
    my $subroutines = _commaify($statistics->subs());
    my $statements = _commaify($statistics->statements_other_than_subs());
    my $lines = _commaify($statistics->lines());
    my $width = max map { length } $files, $subroutines, $statements;

    _out sprintf "%*s %s.\n", $width, $files, 'files';
    _out sprintf "%*s %s.\n", $width, $subroutines, 'subroutines/methods';
    _out sprintf "%*s %s.\n", $width, $statements, 'statements';

    my $lines_of_blank = _commaify( $statistics->lines_of_blank() );
    my $lines_of_comment = _commaify( $statistics->lines_of_comment() );
    my $lines_of_data = _commaify( $statistics->lines_of_data() );
    my $lines_of_perl = _commaify( $statistics->lines_of_perl() );
    my $lines_of_pod = _commaify( $statistics->lines_of_pod() );

    $width =
        max map { length }
            $lines_of_blank, $lines_of_comment, $lines_of_data,
            $lines_of_perl,  $lines_of_pod;
    _out sprintf "\n%s %s:\n",            $lines, 'lines, consisting of';
    _out sprintf "    %*s %s.\n", $width, $lines_of_blank, 'blank lines';
    _out sprintf "    %*s %s.\n", $width, $lines_of_comment, 'comment lines';
    _out sprintf "    %*s %s.\n", $width, $lines_of_data, 'data lines';
    _out sprintf "    %*s %s.\n", $width, $lines_of_perl, 'lines of Perl code';
    _out sprintf "    %*s %s.\n", $width, $lines_of_pod, 'lines of POD';

    my $average_sub_mccabe = $statistics->average_sub_mccabe();
    if (defined $average_sub_mccabe) {
        _out
            sprintf
                "\nAverage McCabe score of subroutines was %.2f.\n",
                $average_sub_mccabe;
        }

    _out "\n";

    _out _commaify($statistics->total_violations()), " violations.\n";

    my $violations_per_file = $statistics->violations_per_file();
    if (defined $violations_per_file) {
        _out
            sprintf
                "Violations per file was %.3f.\n",
                $violations_per_file;
    }
    my $violations_per_statement = $statistics->violations_per_statement();
    if (defined $violations_per_statement) {
        _out
            sprintf
                "Violations per statement was %.3f.\n",
                $violations_per_statement;
    }
    my $violations_per_line = $statistics->violations_per_line_of_code();
    if (defined $violations_per_line) {
        _out
            sprintf
                "Violations per line of code was %.3f.\n",
                $violations_per_line;
    }

    if ( $statistics->total_violations() ) {
        _out "\n";

        my %severity_violations = %{ $statistics->violations_by_severity() };
        my @severities = reverse sort keys %severity_violations;
        $width =
            max
                map { length _commaify( $severity_violations{$_} ) }
                    @severities;
        foreach my $severity (@severities) {
            _out
                sprintf
                    "%*s severity %d violations.\n",
                    $width,
                    _commaify( $severity_violations{$severity} ),
                    $severity;
        }

        _out "\n";

        my %policy_violations = %{ $statistics->violations_by_policy() };
        my @policies = sort keys %policy_violations;
        $width =
            max
                map { length _commaify( $policy_violations{$_} ) }
                    @policies;
        foreach my $policy (@policies) {
            _out
                sprintf
                    "%*s violations of %s.\n",
                    $width,
                    _commaify($policy_violations{$policy}),
                    policy_short_name($policy);
        }
    }

    return;
}

#-----------------------------------------------------------------------------

# Only works for integers.
sub _commaify {
    my ( $number ) = @_;

    while ($number =~ s/ \A ( [-+]? \d+ ) ( \d{3} ) /$1,$2/xms) {
        # nothing
    }

    return $number;
}

#-----------------------------------------------------------------------------

sub _get_option_specification {

    return qw<
        5 4 3 2 1
        Safari
        version
        brutal
        count|C
        cruel
        doc=s
        exclude=s@
        force!
        gentle
        harsh
        help|?|H
        include=s@
        list
        list-enabled
        list-themes
        man
        color|colour!
        noprofile
        only!
        options
        pager=s
        profile|p=s
        profile-proto
        quiet
        severity=i
        single-policy|s=s
        stern
        statistics!
        statistics-only!
        profile-strictness=s
        theme=s
        top:i
        verbose=s
        color-severity-highest|colour-severity-highest|color-severity-5|colour-severity-5=s
        color-severity-high|colour-severity-high|color-severity-4|colour-severity-4=s
        color-severity-medium|colour-severity-medium|color-severity-3|colour-severity-3=s
        color-severity-low|colour-severity-low|color-severity-2|colour-severity-2=s
        color-severity-lowest|colour-severity-lowest|color-severity-1|colour-severity-1=s
        files-with-violations|l
        files-without-violations|L
    >;
}

#-----------------------------------------------------------------------------

sub _colorize_by_severity {
    my @violations = @_;
    return @violations if _this_is_windows();
    return @violations if not eval { require Term::ANSIColor; 1; };

    my $config = $critic->config();
    my %color_of = (
        $SEVERITY_HIGHEST   => $config->color_severity_highest(),
        $SEVERITY_HIGH      => $config->color_severity_high(),
        $SEVERITY_MEDIUM    => $config->color_severity_medium(),
        $SEVERITY_LOW       => $config->color_severity_low(),
        $SEVERITY_LOWEST    => $config->color_severity_lowest(),
    );

    return map { _colorize( "$_", $color_of{$_->severity()} ) } @violations;

}

#-----------------------------------------------------------------------------

sub _colorize {
    my ($string, $color) = @_;
    return $string if not defined $color;
    return $string if $color eq $EMPTY;
    # $terminator is a purely cosmetic change to make the color end at the end
    # of the line rather than right before the next line. It is here because
    # if you use background colors, some console windows display a little
    # fragment of colored background before the next uncolored (or
    # differently-colored) line.
    my $terminator = chomp $string ? "\n" : $EMPTY;
    return  Term::ANSIColor::colored( $string, $color ) . $terminator;
}

#-----------------------------------------------------------------------------

sub _this_is_windows {
    return 1 if $OSNAME =~ m/MSWin32/xms;
    return 0;
}

#-----------------------------------------------------------------------------

sub _at_tty {
    return -t STDOUT; ## no critic (ProhibitInteractiveTest);
}

#-----------------------------------------------------------------------------

sub _render_all_policy_listing {
    # Force P-C parameters, to catch all Policies on this site
    my %pc_params = (-profile => $EMPTY, -severity => $SEVERITY_LOWEST);
    return _render_policy_listing( %pc_params );
}

#-----------------------------------------------------------------------------

sub _render_policy_listing {
    my %pc_params = @_;

    require Perl::Critic::PolicyListing;
    require Perl::Critic;

    my @policies = Perl::Critic->new( %pc_params )->policies();
    my $listing = Perl::Critic::PolicyListing->new( -policies => \@policies );
    _out $listing;

    exit $EXIT_SUCCESS;
}

#-----------------------------------------------------------------------------

sub _render_theme_listing {

    require Perl::Critic::ThemeListing;
    require Perl::Critic;

    my %pc_params = (-profile => $EMPTY, -severity => $SEVERITY_LOWEST);
    my @policies = Perl::Critic->new( %pc_params )->policies();
    my $listing = Perl::Critic::ThemeListing->new( -policies => \@policies );
    _out $listing;

    exit $EXIT_SUCCESS;
}

#-----------------------------------------------------------------------------

sub _render_profile_prototype {

    require Perl::Critic::ProfilePrototype;
    require Perl::Critic;

    my %pc_params = (-profile => $EMPTY, -severity => $SEVERITY_LOWEST);
    my @policies = Perl::Critic->new( %pc_params )->policies();
    my $prototype = Perl::Critic::ProfilePrototype->new( -policies => \@policies );
    _out $prototype;

    exit $EXIT_SUCCESS;
}

#-----------------------------------------------------------------------------

sub _render_policy_docs {

    my (%opts) = @_;
    my $pattern = delete $opts{-doc};

    require Perl::Critic;
    $critic = Perl::Critic->new(%opts);
    _set_up_pager($critic->config()->pager());

    require Perl::Critic::PolicyFactory;
    my @site_policies  = Perl::Critic::PolicyFactory->site_policy_names();
    my @matching_policies  = grep { $_ =~ m/$pattern/ixms } @site_policies;

    # "-T" means don't send to pager
    my @perldoc_output = map {`perldoc -T $_`} @matching_policies;  ## no critic (ProhibitBacktick)
    _out @perldoc_output;

    exit $EXIT_SUCCESS;
}

#-----------------------------------------------------------------------------

sub _display_version {
    _out "$VERSION\n";
    exit $EXIT_SUCCESS;
}

#-----------------------------------------------------------------------------
1;

__END__

#-----------------------------------------------------------------------------

=pod

=for stopwords
Twitter

=head1 NAME

Perl::Critic::Command - Guts of L<perlcritic>.


=head1 SYNOPSIS

    use Perl::Critic::Command qw< run >;

    local @ARGV = qw< --statistics-only lib bin >;
    run();


=head1 DESCRIPTION

This is the implementation of the L<perlcritic> command.  You can use
this to run the command without going through a command interpreter.


=head1 INTERFACE SUPPORT

This is considered to be a public class.  However, its interface is
experimental, and will likely change.


=head1 IMPORTABLE SUBROUTINES

=over

=item C<run()>

Does the equivalent of the L<perlcritic> command.  Unfortunately, at
present, this doesn't take any parameters but uses C<@ARGV> to get its
input instead.  Count on this changing; don't count on the current
interface.


=back


=head1 TO DO

Make C<run()> take parameters.  The equivalent of C<@ARGV> should be
passed as a reference.

Turn this into an object.


=head1 AUTHOR

Jeffrey Ryan Thalhammer <thaljef@cpan.org>


=head1 COPYRIGHT

Copyright (c) 2005-2009 Jeffrey Ryan Thalhammer.  All rights reserved.

This program is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.  The full text of this license
can be found in the LICENSE file included with this module.

=cut

##############################################################################
# Local Variables:
#   mode: cperl
#   cperl-indent-level: 4
#   fill-column: 78
#   indent-tabs-mode: nil
#   c-indentation-style: bsd
# End:
# ex: set ts=8 sts=4 sw=4 tw=78 ft=perl expandtab shiftround :
